src/Uno.UI.RuntimeTests/Tests/Windows_Data/Pdf/Given_PdfDocument.cs

```diff
@@ -0,0 +1,275 @@
+﻿#nullable enable
+#if !HAS_UNO || XAMARIN_ANDROID
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Private.Infrastructure;
+using Uno.UI.RuntimeTests.Helpers;
+using Windows.Data.Pdf;
+using Windows.Storage;
+using Windows.Storage.Streams;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Media.Imaging;
+
+namespace Uno.UI.RuntimeTests.Tests.Windows_Data.Pdf;
+
+[TestClass]
+[RunsOnUIThread]
+#if HAS_UNO && !XAMARIN_ANDROID
+[Ignore("Not implemented yet.")]
+#endif
+public class Given_PdfDocument
+{
+	private const string PdfDocument_Password_Valid = "uno";
+	private const string PdfDocument_Password_Invalid = "1";
+	private const string PdfDocument_Name = "UnoA4.pdf";
+	private const string PdfDocument_Name_Protected = "UnoA4_Protected.pdf";
+	private const string ReferencePageImageUri = "UnoA4.png";
+	private const string ReferencePageImage_ProtectedUri = "UnoA4_Protected.png";
+
+	/// <summary>
+	/// Calculates root mean square error for given two images.
+	/// Based roughly on ImageMagick implementation to ensure consistency.
+	/// If the error is greater than or equal to 0.022, the differences are visible to human eyes.
+	/// </summary>
+	private static double CompareImages(RawBitmap actual, RawBitmap expected)
```

Comment:
There is HasPixels but you can set the tolerance at the pixel level but not the whole image. The images, although different, are identical to the human eye. If the error returned by the function is greater than 0.022, the differences are visible to the human eye.

Reply:
I had thought of moving it to `ImageAssert` but I am not convinced by the name `CompareImages`

Reply:
Yes we do already have such comparison methods here https://github.com/unoplatform/uno/blob/master/src/SamplesApp/SamplesApp.UITests/TestFramework/ImageAssert.cs#L49 They are not yet ported to the RuntimeTests but it would be great to port them :)

Reply:
Hi @dr1rrb ,
I've seen the method you suggested, but it doesn't do exactly what I need.

AreAlmostEqual: verifies for each single pixel of the images to be compared that the sum of the absolute values of the differences between the individual components of the pixels is less than the specified tolerance.

CompareImages: Computes the [root mean square error](https://en.wikipedia.org/wiki/Root-mean-square_deviation) for two given images

Reply:
Yeah I always considered that our comparison of pixels was pretty weak. What would be nice then it to add that comparison type to the [`ColorToleranceKind`](https://github.com/unoplatform/uno/blob/f1b6f49acc01d584939b4dcc3d7dec885fd09ed1/src/SamplesApp/SamplesApp.UITests/TestFramework/ImageAssert.ExpectedPixels.cs#L256) and adjust the comparison to support such comparison (https://github.com/unoplatform/uno/blob/master/src/SamplesApp/SamplesApp.UITests/TestFramework/ImageAssert.Validations.cs#L44) So we would take advantage of it in all UI tests using the same "syntax" to specify comparison regions/rect etc.

But I agree that it might be not be an easy fix and we prefer to do that in another PR, your choice.

(Note: I would assume that the "root mean square error" would not make sense with `LocationToleranceKind.PerPixel`, I would throw an exception if user specified both).

Reply:
Ok, I'll try to do another PR for ImageAssert in the second time. Thanks for the indications.

Reply:
It would still be best to move it to `ImageAssert`, and provide the `ImperceptibilityDeviation` value as a parameter so it can assert directly there. You can also have it call `RawImage.Populate` so it's not called in the test.